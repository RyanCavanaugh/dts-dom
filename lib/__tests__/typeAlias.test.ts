import {create, emit, type} from "../index"

describe("an emitted type alias", () => {
    describe("with a single type parameter", () => {
        it("matches the snapshot", () => {
            const typeParameter = create.typeParameter('T');
            const typeParameterReference = create.namedTypeReference(typeParameter.name)
            const typeAlias = create.alias('test', typeParameterReference);
            typeAlias.typeParameters.push(typeParameter);

            expect(emit(typeAlias)).toMatchSnapshot();
        });

        describe("with a default type", () => {
            it("matches the snapshot", () => {
                const defaultType = type.number;
                const typeParameter = create.typeParameter('T');
                typeParameter.defaultType = defaultType;
                const typeParameterReference = create.namedTypeReference(typeParameter.name)
                const typeAlias = create.alias('test', typeParameterReference);
                typeAlias.typeParameters.push(typeParameter);

                expect(emit(typeAlias)).toMatchSnapshot();
            });
        });

        describe("with a base type", () => {
            it("matches the snapshot", () => {
                const baseType = create.interface('interfc');
                const typeParameter = create.typeParameter('T', baseType);
                const typeParameterReference = create.namedTypeReference(typeParameter.name)
                const typeAlias = create.alias('test', typeParameterReference);
                typeAlias.typeParameters.push(typeParameter);
    
                expect(emit(typeAlias)).toMatchSnapshot();
            });
        });

        describe("with a base type and a default type", () => {
            it("matches the snapshot", () => {
                const baseType = create.interface('interfc');
                const typeParameter = create.typeParameter('T', baseType);
                typeParameter.defaultType = type.any;
                const typeParameterReference = create.namedTypeReference(typeParameter.name)
                const typeAlias = create.alias('test', typeParameterReference);
                typeAlias.typeParameters.push(typeParameter);
    
                expect(emit(typeAlias)).toMatchSnapshot();
            });
        });
    });

    describe("with multiple type parameters", () => {
        it("matches the snapshot", () => {
            const typeParameters = [
                create.typeParameter('T'),
                create.typeParameter('K'),
            ];
            const typeParameterReferences = typeParameters.map(typeParameter => create.namedTypeReference(typeParameter.name));
            const typeAlias = create.alias('test', create.union(typeParameterReferences));
            typeAlias.typeParameters.push(...typeParameters);

            expect(emit(typeAlias)).toMatchSnapshot();
        });

        describe("with a default type", () => {
            it("matches the snapshot", () => {
                const typeParameters = [
                    create.typeParameter('T'),
                    create.typeParameter('K'),
                ];
                typeParameters.forEach(typeParameter => {
                    typeParameter.defaultType = type.object;
                });

                const typeParameterReferences = typeParameters.map(typeParameter => create.namedTypeReference(typeParameter.name));
                const typeAlias = create.alias('test', create.union(typeParameterReferences));
                typeAlias.typeParameters.push(...typeParameters);

                expect(emit(typeAlias)).toMatchSnapshot();
            });
        });

        describe("with a base type", () => {
            it("matches the snapshot", () => {
                const baseType = create.interface('interfc');
                const typeParameters = [
                    create.typeParameter('T', baseType),
                    create.typeParameter('K', baseType),
                ];

                const typeParameterReferences = typeParameters.map(typeParameter => create.namedTypeReference(typeParameter.name));
                const typeAlias = create.alias('test', create.union(typeParameterReferences));
                typeAlias.typeParameters.push(...typeParameters);

                expect(emit(typeAlias)).toMatchSnapshot();
            });
        });

        describe("with a base type and a default type", () => {
            it("matches the snapshot", () => {
                const baseType = create.interface('interfc');
                const typeParameters = [
                    create.typeParameter('T', baseType),
                    create.typeParameter('K', baseType),
                ];
                typeParameters.forEach(typeParameter => {
                    typeParameter.defaultType = type.boolean;
                });

                const typeParameterReferences = typeParameters.map(typeParameter => create.namedTypeReference(typeParameter.name));
                const typeAlias = create.alias('test', create.union(typeParameterReferences));
                typeAlias.typeParameters.push(...typeParameters);

                expect(emit(typeAlias)).toMatchSnapshot();
            });
        });
    });
});
