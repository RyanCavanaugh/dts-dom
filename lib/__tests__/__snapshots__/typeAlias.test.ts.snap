// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`an emitted type alias with a single type parameter matches the snapshot 1`] = `
declare type test<T> = T;


`;

exports[`an emitted type alias with a single type parameter with a base type and a default type matches the snapshot 1`] = `
declare type test<T extends interfc = any> = T;


`;

exports[`an emitted type alias with a single type parameter with a base type matches the snapshot 1`] = `
declare type test<T extends interfc> = T;


`;

exports[`an emitted type alias with a single type parameter with a default type matches the snapshot 1`] = `
declare type test<T = number> = T;


`;

exports[`an emitted type alias with multiple type parameters matches the snapshot 1`] = `
declare type test<T, K> = T | K;


`;

exports[`an emitted type alias with multiple type parameters with a base type and a default type matches the snapshot 1`] = `
declare type test<T extends interfc = boolean, K extends interfc = boolean> = T | K;


`;

exports[`an emitted type alias with multiple type parameters with a base type matches the snapshot 1`] = `
declare type test<T extends interfc, K extends interfc> = T | K;


`;

exports[`an emitted type alias with multiple type parameters with a default type matches the snapshot 1`] = `
declare type test<T = object, K = object> = T | K;


`;
